larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1173 @@
+
+            lookupMap.insert(std::pair<int, double>(n, 0.0));
+            reverseLookupMap.insert(std::pair<double, int>(0.0, n));
+            break;
+        }
+        else
+        {
+            lookupMap.insert(std::pair<int, double>(n, currentEnergy));
+            reverseLookupMap.insert(std::pair<double, int>(currentEnergy, n));
+        }
+
+        currentEnergy -= (currentdEdx * binWidth);
+    }
+
+    //remove redundant entries to make lookup much faster
+    for (std::map<int, double>::iterator it = lookupMap.begin(); it != lookupMap.end(); it++)
+    {
+        double n(it->first);
+        double val(it->second);
+        double distanceFromMaxRange((maxBin - n) * binWidth);
+
+        if (n == 0 || n == maxBin)
+            continue;
+        else
+        {
+            double distanceMagnitude(floor(distanceFromMaxRange / 2.0));
+            double samplingDistance((1 + distanceMagnitude) * binWidth);
+
+            if (!(remainder(distanceFromMaxRange, samplingDistance) == 0.0))
+            {
+                lookupMap.erase(n);
+                reverseLookupMap.erase(val);
+            }
+        }
+    }
+
+    lookupTable.SetMap(lookupMap);
+    lookupTable.SetReverseMap(reverseLookupMap);
+
+    if (lookupTable.GetMaxRange() == 0.f)
+        std::cout << "Warning: the lookup table max range has not been correctly set." << std::endl;
+}
+//----------------------------------------------------------------------------------------------------------------------------------
+
+double TrackDirectionTool::GetEnergyfromLength(LookupTable &lookupTable, double &trackLength)
+{
+    std::map<int, double> lookupMap(lookupTable.GetMap());
+    double binWidth(lookupTable.GetBinWidth());
+
+    if (trackLength >= lookupTable.GetMaxRange())
+        return 0.5; //0 energy means infinite dE/dx
+    else if (trackLength <= 1.0)
+        return lookupTable.GetInitialEnergy();
+
+    int n(std::floor(trackLength / binWidth));
+    std::map<int, double>::iterator nextEntryIterator(lookupMap.upper_bound(n));
+    std::map<int, double>::iterator previousEntryIterator(std::prev(nextEntryIterator, 1));
+
+    double leftLength(previousEntryIterator->first * binWidth), rightLength(nextEntryIterator->first * binWidth);
+    double leftEnergy(previousEntryIterator->second), rightEnergy(nextEntryIterator->second);
+    double lengthDifference(rightLength - leftLength);
+    double energyDifference(leftEnergy - rightEnergy);
+
+    double finalEnergy(leftEnergy - (((trackLength - leftLength) / (lengthDifference)) * (energyDifference)));
+
+    //very small energy values lead to huge dE/dx values: truncate
+    if (finalEnergy <= 2.0)
+        return 2.0;
+    else
+        return finalEnergy;
+}
+
+//----------------------------------------------------------------------------------------------------------------------------------
+
+double TrackDirectionTool::GetLengthfromEnergy(LookupTable &lookupTable, double &currentEnergy)
```

Comment:
`const double &currentEnergy`?  I think the five `double`s defined in this function can be `const` too